dotnet/Utils/RichPath.cs

```diff
@@ -0,0 +1,439 @@
+                    foreach (var element in ReadJSONFromTextReader<List<T>>(textReader))
+                    {
+                        yield return element;
+                    }
+                }
+                else
+                {
+                    throw new ArgumentException($"File suffix must be .json[.gz] or .jsonl[.gz], but got {Path}");
+                }
+            }
+        }
+
+        protected static bool StreamIsGZipped(Stream inStream)
+        {
+            // Read first two bytes as signature, and seek to beginning again:
+            byte[] signature = new byte[2];
+            inStream.Read(signature, 0, 2);
+            inStream.Seek(0, SeekOrigin.Begin);
+            return signature[0] == 0x1f && signature[1] == 0x8b;
+        }
+    }
+
+    public class LocalPath : RichPath
+    {
+        public override string Path { get; }
+
+        public LocalPath(string path)
+        {
+            Path = path;
+        }
+
+        public override string ToString() => Path;
+
+        public override bool IsDirectory() => Directory.Exists(Path);
+
+        public override bool IsFile() => File.Exists(Path);
+
+        public override void MakeAsDirectory() => Directory.CreateDirectory(Path);
+
+        public override Stream ReadAsBinaryPotentiallyUncompress()
+        {
+            var inStream = File.OpenRead(Path);
+
+            if (StreamIsGZipped(inStream))
+            {
+                return new GZipStream(inStream, CompressionMode.Decompress);
+            }
+
+            return inStream;
+        }
+
+        public override void Store(object data)
+        {
+            using (var outStream = new FileStream(Path, FileMode.Create))
+            {
+                string normalizedPath;
+                Stream possiblyCompressedStream;
+                if (Path.EndsWith(".gz"))
+                {
+                    normalizedPath = Path.Substring(0, Path.Length - 3); // Remove the ".gz"
+                    possiblyCompressedStream = new GZipStream(outStream, CompressionMode.Compress);
+                }
+                else
+                {
+                    normalizedPath = Path;
+                    possiblyCompressedStream = outStream;
+                }
+
+                if (normalizedPath.EndsWith(".json"))
+                {
+                    StoreAsJSON(possiblyCompressedStream, data, jsonl: false);
+                }
+                else if (normalizedPath.EndsWith(".jsonl"))
+                {
+                    StoreAsJSON(possiblyCompressedStream, data, jsonl: true);
+                }
+                else
+                {
+                    throw new ArgumentException($"File suffix must be .json[.gz] or .jsonl[.gz], but got {Path}");
+                }
+            }
+        }
+
+        public override IEnumerable<RichPath> FilesInDirectory(string searchPattern = "*")
+        {
+            foreach (var newPath in Directory.EnumerateFiles(Path, searchPattern))
```

Comment:
I think to be similar to the Python (and Azure) implementation this needs to be recursive:
```suggestion
            foreach (var newPath in Directory.EnumerateFiles(Path, searchPattern, SearchOption.AllDirectories))
```

Reply:
Done.